cmd_test.go

```diff
@@ -0,0 +1,432 @@
+package main_test
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"os/exec"
+	"os/user"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"testing"
+	"time"
+
+	lfstransfer "github.com/charmbracelet/git-lfs-transfer"
+	"github.com/go-git/go-git/v5"
+	"github.com/stretchr/testify/assert"
+)
+
+func newTestRepo(tb testing.TB) (*git.Repository, string) {
+	tb.Helper()
+	path := filepath.Join(tb.TempDir(), "repo.git")
+	r, err := git.PlainInit(path, true)
+	if err != nil {
+		tb.Fatal(err)
+	}
+	return r, path
+}
+
+func replaceUserId(s string) string {
+	usr, err := user.Current()
+	if err != nil {
+		return s
+	}
+
+	username := usr.Username
+	switch runtime.GOOS {
+	case "windows":
+		username = "unknown"
+	}
+	s = strings.ReplaceAll(s, "0018ownername=test user\n", fmt.Sprintf("%04xownername=%s\n",
+		4+len("ownername=")+len(username+"\n"),
+		username))
+
+	s = strings.ReplaceAll(s, "0059ownername d76670443f4d5ecdeea34c12793917498e18e858c6f74cd38c4b794273bb5e28 test user\n",
+		fmt.Sprintf("%04xownername d76670443f4d5ecdeea34c12793917498e18e858c6f74cd38c4b794273bb5e28 %s\n",
+			4+len("ownername d76670443f4d5ecdeea34c12793917498e18e858c6f74cd38c4b794273bb5e28 ")+len(username+"\n"),
+			username))
+
+	return s
+}
+
+func TestFailedVerify(t *testing.T) {
+	_, path := newTestRepo(t)
+	msg := strings.Join(
+		[]string{
+			"000eversion 1",
```

Comment:
Check out the tests that I [wrote](https://gitlab.com/gitlab-org/gitlab-shell/-/merge_requests/855/diffs#28d6846be20763b3addf0066d6a1bbd573e3d48d) for GitLab. I created a number of helper functions on top of the pktline library to abstract away the messy protocol details of requests and responses.